bin/ch-run.c

```diff
@@ -113,26 +153,40 @@ int main(int argc, char * argv[])
    args.container_gid = getegid();
    args.container_uid = geteuid();
    args.initial_working_dir = NULL;
+   args.join = false;
+   args.join_ct = 0;
+   args.join_tag = NULL;
    args.private_home = false;
    args.private_tmp = false;
-   args.verbose = 0;
+   verbose = 1;  // in charliecloud.c
    Z_ (setenv("ARGP_HELP_FMT", "opt-doc-col=25,no-dup-args-note", 0));
    Z_ (argp_parse(&argp, argc, argv, 0, &(args.user_cmd_start), &args));
    Te (args.user_cmd_start < argc - 1, "NEWROOT and/or CMD not specified");
    assert(args.binds[USER_BINDS_MAX].src == NULL);  // overrun in argp_parse?
    args.newroot = realpath(argv[args.user_cmd_start++], NULL);
    Tf (args.newroot != NULL, "couldn't resolve image path");
-
-   if (args.verbose) {
-      fprintf(stderr, "newroot: %s\n", args.newroot);
-      fprintf(stderr, "container uid: %u\n", args.container_uid);
-      fprintf(stderr, "container gid: %u\n", args.container_gid);
-      fprintf(stderr, "private /tmp: %d\n", args.private_tmp);
+   if (args.join) {
+      args.join_ct = join_ct();
+      args.join_tag = join_tag();
    }
 
-   setup_namespaces(args.container_uid, args.container_gid);
-   enter_udss(args.newroot, args.writable, args.binds,
-              args.private_tmp, args.private_home);
+   INFO("verbosity: %d", verbose);
+   INFO("newroot: %s", args.newroot);
+   INFO("container uid: %u", args.container_uid);
+   INFO("container gid: %u", args.container_gid);
+   INFO("join: %d %d %s", args.join, args.join_ct, args.join_tag);
+   INFO("private /tmp: %d", args.private_tmp);
+
+   if (args.join)
```

Comment:
this series of steps looks correct as is, but are signal handlers in place here protect against signals interfering with the procedure/result in uncleaned up resources?

Also,  are we confident that if things go south and the job is killed, that if charlie cloud is running under SLURM, that these semaphore and shm files will be cleaned up as part of job cleanup?

Reply:
> this series of steps looks correct as is, but are signal handlers in place here protect against signals interfering with the procedure/result in uncleaned up resources?

No, there aren't any signal handlers. Which signals would pose a risk?

> Also, are we confident that if things go south and the job is killed, that if charlie cloud is running under SLURM, that these semaphore and shm files will be cleaned up as part of job cleanup?

On our systems, yes we are confident.

If they're not cleaned up, though, and there's a collision later: if it's a different user, their job will fail with permission denied (because other users can't access the files), and if it's the same user, something messy will happen and the `--join` coordination will fail somehow.